File: backend/app/schemas/silo_spawn_requests.py
"""Schemas for silo spawn request APIs."""

from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import field_validator, model_validator
from sqlmodel import SQLModel

SiloSpawnRequestStatus = Literal[
    "requested",
    "planned",
    "spawning",
    "running",
    "materialized",
    "failed",
    "cancelled",
]
SiloSpawnRequestScope = Literal["organization", "board", "silo"]
SiloSpawnRequestPriority = Literal["low", "normal", "high", "urgent"]


class SiloSpawnRequestBase(SQLModel):
    display_name: str
    silo_kind: str = "agent"
    scope: SiloSpawnRequestScope = "organization"
    priority: SiloSpawnRequestPriority = "normal"
    board_id: UUID | None = None
    parent_silo_id: UUID | None = None
    source_task_id: UUID | None = None
    desired_role: str | None = None
    source_task_title: str | None = None
    runtime_preference: str | None = None
    summary: str | None = None
    desired_state: dict[str, object] | None = None

    @field_validator(
        "display_name",
        "silo_kind",
        "desired_role",
        "runtime_preference",
        "summary",
        mode="before",
    )
    @classmethod
    def _normalize_text(cls, value: object) -> object:
        if isinstance(value, str):
            trimmed = value.strip()
            return trimmed or None
        return value

    @model_validator(mode="after")
    def _validate_scope(self) -> "SiloSpawnRequestBase":
        if self.display_name is None:
            raise ValueError("display_name is required")
        if self.scope == "organization":
            return self
        if self.scope == "board" and self.board_id is None:
            raise ValueError("board_id is required for board-scoped silo requests")
        if self.scope == "silo" and self.parent_silo_id is None:
            raise ValueError("parent_silo_id is required for silo-scoped silo requests")
        if self.source_task_id is not None and self.scope != "board":
            raise ValueError("source_task_id requires a board-scoped silo request")
        return self


class SiloSpawnRequestCreate(SiloSpawnRequestBase):
    pass


class SiloSpawnRequestUpdate(SQLModel):
    status: SiloSpawnRequestStatus | None = None
    priority: SiloSpawnRequestPriority | None = None
    source_task_title: str | None = None
    summary: str | None = None
    desired_state: dict[str, object] | None = None

    @field_validator("source_task_title", "summary", mode="before")
    @classmethod
    def _normalize_summary(cls, value: object) -> object:
        if isinstance(value, str):
            trimmed = value.strip()
            return trimmed or None
        return value


class SiloSpawnRequestRead(SiloSpawnRequestBase):
    id: UUID
    organization_id: UUID
    requested_by_user_id: UUID | None = None
    source_task_id: UUID | None = None
    source_task_title: str | None = None
    materialized_silo_id: UUID | None = None
    materialized_silo_slug: str | None = None
    materialized_at: datetime | None = None
    slug: str
    status: SiloSpawnRequestStatus
    created_at: datetime
    updated_at: datetime

File: frontend/src/lib/silo-spawn-requests.ts
"use client";

import type { BoardRead } from "@/api/generated/model/boardRead";
import { customFetch } from "@/api/mutator";

export type SiloSpawnRequestScope = "organization" | "board" | "silo";
export type SiloSpawnRequestStatus =
  | "requested"
  | "planned"
  | "spawning"
  | "running"
  | "materialized"
  | "failed"
  | "cancelled";
export type SiloSpawnRequestPriority = "low" | "normal" | "high" | "urgent";

export type SiloSpawnRequest = {
  id: string;
  organization_id: string;
  requested_by_user_id: string | null;
  source_task_id: string | null;
  source_task_title: string | null;
  materialized_silo_id: string | null;
  materialized_silo_slug: string | null;
  materialized_at: string | null;
  slug: string;
  display_name: string;
  silo_kind: string;
  scope: SiloSpawnRequestScope;
  priority: SiloSpawnRequestPriority;
  board_id: string | null;
  parent_silo_id: string | null;
  desired_role: string | null;
  runtime_preference: string | null;
  status: SiloSpawnRequestStatus;
  summary: string | null;
  desired_state: Record<string, unknown> | null;
  created_at: string;
  updated_at: string;
};

export async function fetchSiloSpawnRequests(): Promise<SiloSpawnRequest[]> {
  const response = await customFetch<{ data: SiloSpawnRequest[] }>("/api/v1/silos/spawn-requests", {
    method: "GET",
  });
  return response.data;
}

export async function fetchSiloSpawnRequestsForBoard(
  boardId: string,
): Promise<SiloSpawnRequest[]> {
  const response = await customFetch<{ data: SiloSpawnRequest[] }>(
    `/api/v1/silos/spawn-requests?board_id=${encodeURIComponent(boardId)}`,
    {
      method: "GET",
    },
  );
  return response.data;
}

export async function createSiloSpawnRequest(payload: {
  display_name: string;
  silo_kind: string;
  scope: SiloSpawnRequestScope;
  priority?: SiloSpawnRequestPriority;
  board_id?: string | null;
  parent_silo_id?: string | null;
  source_task_id?: string | null;
  desired_role?: string | null;
  source_task_title?: string | null;
  runtime_preference?: string | null;
  summary?: string | null;
}): Promise<SiloSpawnRequest> {
  const response = await customFetch<{ data: SiloSpawnRequest }>(
    "/api/v1/silos/spawn-requests",
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
  return response.data;
}

export async function fetchSiloSpawnRequest(requestId: string): Promise<SiloSpawnRequest> {
  const response = await customFetch<{ data: SiloSpawnRequest }>(
    `/api/v1/silos/spawn-requests/${requestId}`,
    {
      method: "GET",
    },
  );
  return response.data;
}

export async function updateSiloSpawnRequest(
  requestId: string,
  payload: {
    status?: SiloSpawnRequestStatus;
    priority?: SiloSpawnRequestPriority;
    source_task_title?: string | null;
    summary?: string | null;
    desired_state?: Record<string, unknown> | null;
  },
): Promise<SiloSpawnRequest> {
  const response = await customFetch<{ data: SiloSpawnRequest }>(
    `/api/v1/silos/spawn-requests/${requestId}`,
    {
      method: "PATCH",
      body: JSON.stringify(payload),
    },
  );
  return response.data;
}

export async function fetchBoardOptions(): Promise<BoardRead[]> {
  const response = await customFetch<{ data: { items: BoardRead[] } }>("/api/v1/boards?limit=200&offset=0", {
    method: "GET",
  });
  return response.data.items;
}
